src/librustc_metadata/rmeta/decoder.rs

```diff
@@ -1343,6 +1343,8 @@ impl<'a, 'tcx> CrateMetadata {
         let constness = match self.kind(id) {
             EntryKind::Method(data) => data.decode(self).fn_data.constness,
             EntryKind::Fn(data) => data.decode(self).constness,
+            // Some intrinsics can be const fn
+            EntryKind::ForeignFn(data) => data.decode(self).constness,
             EntryKind::Variant(..) | EntryKind::Struct(..) => hir::Constness::Const,
             _ => hir::Constness::NotConst,
         };
```

Comment:
Is it expected that the logic here partially duplicates `is_const_fn_raw` in `constness.rs`?

Reply:
kind of. This is for when the `DefId` is not local and we need to load it from metadata. Alternatively we could recompute depending on the signature instead of encoding the constness in metadata.

Reply:
The implementation seems fine, but more comments would be good.

Reply:
(This concern is still open.)

Reply:
@oli-obk there's still a comment missing here.